repository/src/lib.rs

```diff
@@ -156,13 +156,184 @@ impl<T: RawRepository> DistributedRepository<T> {
+            let branch_tip_semantic_commit = self
+                .raw
+                .read_semantic_commit(branch_tip_commit_hash)
+                .await?;
+            new_semantic_commits.push((branch_tip_semantic_commit, branch_tip_commit_hash));
+            let new_commits = new_semantic_commits
+                .into_iter()
+                .map(|(commit, hash)| {
+                    from_semantic_commit(commit)
+                        .map_err(|e| (e, hash))
+                        .map(|x| (x, hash))
+                })
+                .collect::<Result<Vec<_>, _>>()
+                .map_err(|(error, hash)| {
+                    anyhow!("failed to convert the commit {}: {}", hash, error)
+                })?;
+
+            // Verify all the incoming commits and apply them to the local repository only if they are valid.
+            let mut verifier =
+                CommitSequenceVerifier::new(last_header.clone(), reserved_state.clone())
+                    .map_err(|e| anyhow!("failed to create a verifier: {}", e))?;
+            for (new_commit, _) in &new_commits {
+                if verifier.apply_commit(new_commit).is_err() {
+                    continue 'branch;
+                }
+            }
+            let branch_tip_commit = &new_commits.last().unwrap().0;
+            if let Commit::Agenda(_) = branch_tip_commit {
+                // If branch tip commit is Agenda commit, create a new agenda branch
+                let branch_tip_commit_hash = self.raw.locate_branch(branch).await?;
+                let branch_name = "a-"; //TODO : change # into number
+                let branch_name = format!("{}{}", branch_name, branch_tip_commit_hash);
+                self.raw
+                    .create_branch(branch_name, branch_tip_commit_hash)
+                    .await?;
+            } else if let Commit::AgendaProof(_) = branch_tip_commit {
+                // Else if branch tip commit is AgendaProof commit, create a new agenda branch
+                let branch_tip_commit_hash = self.raw.locate_branch(branch).await?;
+                let branch_name = "a-#"; //TODO : change # into number
+                let branch_name = format!("{}{}", branch_name, branch_tip_commit_hash);
+                self.raw
+                    .create_branch(branch_name, branch_tip_commit_hash)
+                    .await?;
+            } else if let Commit::Block(block_header) = branch_tip_commit {
+                // find fp, if there is fp then move finalized branch
+                let fp_commit_hash = self.raw.locate_branch(FP_BRANCH_NAME.into()).await?;
+                let fp_semantic_commit = self.raw.read_semantic_commit(fp_commit_hash).await?;
+                let finalization_proof: FinalizationProof =
+                    serde_json::from_str(&fp_semantic_commit.body).unwrap();
+                let result = verify::verify_finalization_proof(
+                    block_header_vec.get(0).unwrap(),
```

Comment:
when this line is first met, can't the block_header vector be empty?

Reply:
also why 0 index?

Reply:
```suggestion
                    block_header
```
maybe this was your intention?

Reply:
Yes! fixed.